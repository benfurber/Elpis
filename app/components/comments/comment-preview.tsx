import React, { Component } from "react";
import { StyleSheet, Text, View } from "react-native";

import ActionSheet from "react-native-actionsheet";
import { TouchableOpacity } from "react-native-gesture-handler";

import {
  AuthorInfo,
  Badge,
  IconDiscussionLevel,
  Icon,
  ActionSheetComment,
} from "components";
import { Comment as CommentInterface, NavigationType, Post } from "interfaces";
import { colours, elements, layout } from "styles";

interface Props {
  backToText?: string;
  item: CommentInterface;
  navigation: NavigationType;
  postId: Post["id"];
  totalComments: number;
}

class CommentPreview extends Component<Props> {
  ActionSheet = ActionSheet;

  displayContent() {
    const { content } = this.props.item;
    const characterLimit = 250;

    if (!content || content.length < characterLimit) return content;

    return content.substring(0, characterLimit) + "...";
  }

  pressActionIcon() {
    const { totalReplies } = this.props.item;

    if (totalReplies > 0) {
      return <Badge left={0} number={totalReplies} staticPosition />;
    }

    return (
      <Icon
        colour={colours.navyBlueDarkTransparentHigh}
        name="angle-double-right"
      />
    );
  }

  onLongPress() {
    const { isAuthorCurrentUser } = this.props.item;

    if (isAuthorCurrentUser) {
      this.ActionSheet.show();
    }
  }

  render() {
    const { backToText, item, navigation, postId, totalComments } = this.props;
    const { discussionLevel, id } = item;

    const onPress = () =>
      navigation.navigate("Reply", {
        backToText,
        id,
        isFromTopic: true,
        postId,
        totalComments,
      });

    return (
      <TouchableOpacity
        onPress={onPress}
        onLongPress={() => this.onLongPress()}
      >
        <View style={styles.commentContainer}>
          <View style={styles.headerRow}>
            <View style={styles.titleContainer}>
              <AuthorInfo navigation={navigation} item={item} />
            </View>
            <View style={styles.categoryContainer}>
              <View style={styles.categoryIconContainer}>
                <IconDiscussionLevel
                  containerStyle={{ borderBottomRightRadius: 0 }}
                  level={discussionLevel}
                />
              </View>
            </View>
          </View>
          <View style={styles.contentContainer}>
            <Text style={styles.text}>{this.displayContent()}</Text>
            <View style={styles.iconContainer}>{this.pressActionIcon()}</View>
          </View>
        </View>
        <ActionSheetComment
          item={item}
          navigation={navigation}
          postId={postId}
          refProp={o => (this.ActionSheet = o)}
        />
      </TouchableOpacity>
    );
  }
}

const styles = StyleSheet.create({
  avatarContainer: {
    alignItems: "center",
    flexDirection: "row",
    width: 60,
  },
  categoryContainer: {
    alignItems: "flex-end",
    width: 40,
  },
  categoryIconContainer: {
    borderBottomRightRadius: 0,
    borderColor: colours.pureWhite,
    borderRadius: 30,
    borderWidth: 5,
    overflow: "hidden",
    position: "absolute",
    right: -10,
    top: -30,
    zIndex: 1,
  },
  commentContainer: {
    alignItems: "stretch",
    backgroundColor: colours.pureWhite,
    borderRadius: layout.borderRadiusL,
    flex: 1,
    marginBottom: layout.spacingL,
    marginLeft: layout.spacing,
    marginTop: 20,
    padding: layout.spacing,
  },
  contentContainer: {
    flex: 1,
    flexDirection: "row",
    marginVertical: layout.spacing,
  },
  headerRow: {
    flexDirection: "row",
    justifyContent: "space-between",
  },
  iconContainer: {
    alignItems: "center",
    flexDirection: "row",
  },
  text: {
    flex: 1,
    flexWrap: "nowrap",
    marginRight: layout.spacingXL,
  },
  titleContainer: {
    flex: 1,
  },
});

export { CommentPreview };
