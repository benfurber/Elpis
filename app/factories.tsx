import { Comment, Community, Reply, User } from "interfaces";

export const community: Community = {
  avatarPath: require("assets/images/empower_two_women_logo.png"),
  id: "1235we",
  name: "Empodere Duas Mulheres",
  posts: [],
};

export const user: User = {
  avatarPath: require("assets/images/profile-pic-annon.png"),
  id: "1235we",
  name: "May",
};

export const post = {
  author: community,
  comments: [],
  content: null,
  id: "10387-314fs-12asdbj",
  imagePath: require("assets/images/image_post_1.jpg"),
  publishedAt: new Date("2000-01-01"),
};

export const preComment: Comment = {
  author: user,
  content:
    "Dá certo sim, o meu pai, por exemplo, fugiu quando eu tinha 5 anos e eu não faço ideia da onde ele esteja. Nunca mais voltou, pena que a violência só piorou.",
  discussionLevel: 0,
  edited: false,
  id: "21097",
  isAuthorCurrentUser: false,
  publishedAt: new Date("2019-01-01"),
  replies: [],
  totalReplies: 0,
};

export const reply: Reply = {
  author: user,
  comment: preComment,
  content: "Nao.",
  edited: false,
  id: "21097",
  isAuthorCurrentUser: false,
  link: null,
  publishedAt: new Date("2019-01-01"),
};

export const comment: Comment = {
  author: user,
  content:
    "Dá certo sim, o meu pai, por exemplo, fugiu quando eu tinha 5 anos e eu não faço ideia da onde ele esteja. Nunca mais voltou, pena que a violência só piorou.",
  discussionLevel: 0,
  edited: false,
  id: "21097",
  isAuthorCurrentUser: false,
  publishedAt: new Date("2019-01-01"),
  replies: [reply],
  title: "Meu pai fez o que ela mandou",
  totalReplies: 1,
};
