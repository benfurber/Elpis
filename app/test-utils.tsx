export const mockDateNow = date => {
  const dateToday = new Date(date);
  jest.spyOn(Date, "now").mockImplementation(() => dateToday.getTime());
};
