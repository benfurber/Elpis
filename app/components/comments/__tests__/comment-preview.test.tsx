import "react-native";
import React from "react";
import { shallow } from "enzyme";

import { author, comment } from "../../../factories";
import { mockDateNow } from "../../../test-utils";

import { CommentPreview } from "../comment-preview";

let navigation;

describe("Comments<CommentPreview>", () => {
  describe("with replies", () => {
    it("renders correctly", () => {
      mockDateNow("2019-01-01T20:04:23");

      const onPress = () => jest.fn();
      const postId = "213543";

      const component = shallow(
        <CommentPreview
          item={comment}
          onPress={onPress}
          navigation={navigation}
          postId={postId}
        />,
      );

      expect(component).toMatchSnapshot();
    });
  });

  describe("without replies", () => {
    it("renders correctly", () => {
      mockDateNow("2019-01-01T00:09:00");

      const comment = {
        author,
        content:
          "Dá certo sim, o meu pai, por exemplo, fugiu quando eu tinha 5 anos e eu não faço ideia da onde ele esteja. Nunca mais voltou, pena que a violência só piorou.",
        id: "21097",
        isAuthorCurrentUser: false,
        publishedAt: new Date("2019-01-01"),
        replies: [],
        totalReplies: 0,
      };
      const onPress = () => jest.fn();
      const postId = "927164";

      const component = shallow(
        <CommentPreview
          item={comment}
          onPress={onPress}
          navigation={navigation}
          postId={postId}
        />,
      );

      expect(component).toMatchSnapshot();
    });
  });
});
