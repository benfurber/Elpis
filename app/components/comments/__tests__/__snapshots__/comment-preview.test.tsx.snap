// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comments<CommentPreview> with replies renders correctly 1`] = `
<TouchableOpacity
  activeOpacity={0.2}
  delayLongPress={600}
  extraButtonProps={
    Object {
      "rippleColor": "transparent",
    }
  }
  onLongPress={[Function]}
  onPress={[Function]}
>
  <View
    style={
      Object {
        "alignItems": "stretch",
        "borderTopColor": "rgb(255, 255, 255)",
        "borderTopWidth": 3,
        "flex": 1,
        "flexDirection": "row",
        "margin": 10,
        "marginBottom": 0,
        "paddingTop": 5,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "width": 60,
        }
      }
    >
      <View>
        <Avatar
          avatarPath={
            Object {
              "testUri": "../../../app/assets/images/empower_two_women_logo.png",
            }
          }
          userId="1235we"
        />
      </View>
    </View>
    <View
      style={
        Object {
          "flex": 1,
          "marginLeft": 10,
          "marginVertical": 10,
        }
      }
    >
      <Title
        text="Meu pai fez o que ela mandou"
      />
      <Text
        style={
          Object {
            "color": "rgb(51, 51, 51)",
            "fontStyle": "italic",
            "marginBottom": 10,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "rgb(45, 82, 113)",
              "fontWeight": "bold",
            }
          }
        >
          Empodere Duas Mulheres 
        </Text>
        há 20 horas
      </Text>
      <Text
        style={
          Object {
            "flex": 1,
            "flexWrap": "nowrap",
          }
        }
      >
        Dá certo sim, o meu pai, por exemplo, fugiu quando eu tinha 5 anos e...
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "marginRight": 3.3333333333333335,
        }
      }
    >
      <Badge
        left={0}
        number={1}
        staticPosition={true}
      />
    </View>
  </View>
  <ActionSheetComment
    item={
      Object {
        "author": Object {
          "avatarPath": Object {
            "testUri": "../../../app/assets/images/empower_two_women_logo.png",
          },
          "id": "1235we",
          "name": "Empodere Duas Mulheres",
          "posts": Array [],
        },
        "content": "Dá certo sim, o meu pai, por exemplo, fugiu quando eu tinha 5 anos e eu não faço ideia da onde ele esteja. Nunca mais voltou, pena que a violência só piorou.",
        "edited": false,
        "id": "21097",
        "isAuthorCurrentUser": false,
        "publishedAt": 2019-01-01T00:00:00.000Z,
        "replies": Array [
          Object {
            "author": Object {
              "avatarPath": Object {
                "testUri": "../../../app/assets/images/empower_two_women_logo.png",
              },
              "id": "1235we",
              "name": "Empodere Duas Mulheres",
              "posts": Array [],
            },
            "comment": Object {
              "author": Object {
                "avatarPath": Object {
                  "testUri": "../../../app/assets/images/empower_two_women_logo.png",
                },
                "id": "1235we",
                "name": "Empodere Duas Mulheres",
                "posts": Array [],
              },
              "content": "Dá certo sim, o meu pai, por exemplo, fugiu quando eu tinha 5 anos e eu não faço ideia da onde ele esteja. Nunca mais voltou, pena que a violência só piorou.",
              "edited": false,
              "id": "21097",
              "isAuthorCurrentUser": false,
              "publishedAt": 2019-01-01T00:00:00.000Z,
              "replies": Array [],
              "totalReplies": 0,
            },
            "content": "Nao.",
            "edited": false,
            "id": "21097",
            "isAuthorCurrentUser": false,
            "link": null,
            "publishedAt": 2019-01-01T00:00:00.000Z,
          },
        ],
        "title": "Meu pai fez o que ela mandou",
        "totalReplies": 1,
      }
    }
    postId="213543"
    refProp={[Function]}
  />
</TouchableOpacity>
`;

exports[`Comments<CommentPreview> without replies renders correctly 1`] = `
<TouchableOpacity
  activeOpacity={0.2}
  delayLongPress={600}
  extraButtonProps={
    Object {
      "rippleColor": "transparent",
    }
  }
  onLongPress={[Function]}
  onPress={[Function]}
>
  <View
    style={
      Object {
        "alignItems": "stretch",
        "borderTopColor": "rgb(255, 255, 255)",
        "borderTopWidth": 3,
        "flex": 1,
        "flexDirection": "row",
        "margin": 10,
        "marginBottom": 0,
        "paddingTop": 5,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "width": 60,
        }
      }
    >
      <View>
        <Avatar
          avatarPath={
            Object {
              "testUri": "../../../app/assets/images/empower_two_women_logo.png",
            }
          }
          userId="1235we"
        />
      </View>
    </View>
    <View
      style={
        Object {
          "flex": 1,
          "marginLeft": 10,
          "marginVertical": 10,
        }
      }
    >
      <Title
        text="Meu pai fez o que ela mandou"
      />
      <Text
        style={
          Object {
            "color": "rgb(51, 51, 51)",
            "fontStyle": "italic",
            "marginBottom": 10,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "rgb(45, 82, 113)",
              "fontWeight": "bold",
            }
          }
        >
          Empodere Duas Mulheres 
        </Text>
        há 9 minutos
      </Text>
      <Text
        style={
          Object {
            "flex": 1,
            "flexWrap": "nowrap",
          }
        }
      >
        Dá certo sim, o meu pai, por exemplo, fugiu quando eu tinha 5 anos e...
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "marginRight": 3.3333333333333335,
        }
      }
    >
      <Badge
        left={0}
        number={1}
        staticPosition={true}
      />
    </View>
  </View>
  <ActionSheetComment
    item={
      Object {
        "author": Object {
          "avatarPath": Object {
            "testUri": "../../../app/assets/images/empower_two_women_logo.png",
          },
          "id": "1235we",
          "name": "Empodere Duas Mulheres",
          "posts": Array [],
        },
        "content": "Dá certo sim, o meu pai, por exemplo, fugiu quando eu tinha 5 anos e eu não faço ideia da onde ele esteja. Nunca mais voltou, pena que a violência só piorou.",
        "edited": false,
        "id": "21097",
        "isAuthorCurrentUser": false,
        "publishedAt": 2019-01-01T00:00:00.000Z,
        "replies": Array [
          Object {
            "author": Object {
              "avatarPath": Object {
                "testUri": "../../../app/assets/images/empower_two_women_logo.png",
              },
              "id": "1235we",
              "name": "Empodere Duas Mulheres",
              "posts": Array [],
            },
            "comment": Object {
              "author": Object {
                "avatarPath": Object {
                  "testUri": "../../../app/assets/images/empower_two_women_logo.png",
                },
                "id": "1235we",
                "name": "Empodere Duas Mulheres",
                "posts": Array [],
              },
              "content": "Dá certo sim, o meu pai, por exemplo, fugiu quando eu tinha 5 anos e eu não faço ideia da onde ele esteja. Nunca mais voltou, pena que a violência só piorou.",
              "edited": false,
              "id": "21097",
              "isAuthorCurrentUser": false,
              "publishedAt": 2019-01-01T00:00:00.000Z,
              "replies": Array [],
              "totalReplies": 0,
            },
            "content": "Nao.",
            "edited": false,
            "id": "21097",
            "isAuthorCurrentUser": false,
            "link": null,
            "publishedAt": 2019-01-01T00:00:00.000Z,
          },
        ],
        "title": "Meu pai fez o que ela mandou",
        "totalReplies": 1,
      }
    }
    postId="927164"
    refProp={[Function]}
  />
</TouchableOpacity>
`;
